### PR TITLE
Add AST node for variable declarators for better quality debug info

### PR DESCRIPTION
**Pull request by qingyuanzNV**
At the moment, GLSLANG drops variable declaration info immediately after making the symbol in the symbol table. This leads to difficulty in emitting debug info, presenting that 1) debug info for constant variable is missing 2) debug line of variables points to its first use instead of declaration.

This patch introduces a new `TIntermVariableDecl` node in the AST, consisting of an `TIntermSymbol* declSymbol` and an `TIntermNode* initNode`. `declSymbol` is an imaginary access to the declared variable, being used to instantiate the basic variable info and debug information at the point of declaration. It doesn't participate AST traversal, such that code generation is still operating on the same AST as before. `TIntermNode* initNode` is the AST node for the initializer, which used to be emitted directly in the OpSequence. It's now wrapped within a TIntermVariableDecl node. For example,

- When a declaration "int ^x;" is seen, a new tree node is inserted to the AST pointing to ^.
- When a declaration "int ^x = 42;" is seen, the initializer node in the AST is replace with the new tree node pointing to ^.

### CODE REVIEW

File: glslang/MachineIndependent/ParseHelper.cpp [hunk 1]
```diff
@@ -10323,6 +10338,17 @@ void TParseContext::declareBlock(const TSourceLoc& loc, TTypeList& typeList, con
 
     // Save it in the AST for linker use.
     trackLinkage(variable);
+
+    TIntermAggregate* declNode = nullptr;
+    if (intermediate.getDebugInfo()) {
+        auto blockDeclNode = new TIntermVariableDecl(intermediate.addSymbol(variable, loc), nullptr);
+        blockDeclNode->setLoc(loc);
+
+        // We have to wrap the decalaration with a sequence to fit the same processing logic with variables.
```
Comment (dnovillo) [inline glslang/MachineIndependent/ParseHelper.cpp]: nit: s/decalaration/declaration/

File: glslang/Include/intermediate.h [hunk 2]
```diff
@@ -1092,6 +1095,37 @@ struct TIntermNodePair {
...
+
+    TIntermSymbol* getDeclSymbol() { return declSymbol; }
+    const TIntermSymbol* getDeclSymbol() const { return declSymbol; }
+
+    TIntermNode* getInitNode() { return initNode; }
+    const TIntermNode* getInitNode() const { return initNode; }
+
+private:
+    // This symbol is an imaginary access of the variable defined, which isn't an AST node and 
+    // doesn't participate tree traversal.
```
Comment (dnovillo) [inline glslang/Include/intermediate.h]: nit: s/participate tree/participate in tree/

**Comment by qingyuanzNV** [comment]
> Some initial comments (still going through it). Is it feasible to add tests for some edge cases like re-declarations, nested scopes with multiple declarations and arrays? I don't think I've seen those.

Done.

> I'm still trying to understand the different handling for declarations in the grammar. regular variables, block declarations and condition declarations emit different patterns. Do we need grammar changes here? It doesn't seem like we do, but I think I'm missing something.

Yes, we have grammar change in glslang.y and glslang_tab.cpp

---

I have rebased and and modified the traverser to visit the imaginary symbol if wanted. This makes the corner case to be handled more elegantly. @dnovillo please help review.

File: glslang/MachineIndependent/glslang.y [hunk 3]
```diff
@@ -3990,11 +3990,20 @@ condition
...
-        TIntermNode* initNode = parseContext.declareVariable($2.loc, *$2.string, $1, 0, $4);
-        if (initNode)
-            $$ = initNode->getAsTyped();
-        else
+        TIntermNode* declNode = parseContext.declareVariable($2.loc, *$2.string, $1, 0, $4);
+        if (declNode) {
+            if (declNode->getAsVariableDecl()) {
+                // FIXME: could we support decl AST in this case?
+                $$ = declNode->getAsVariableDecl()->getInitNode()->getAsTyped();
+                delete declNode;
```
Comment (dnovillo) [inline glslang/MachineIndependent/glslang.y]: I'm concerned about double-free issues or memory corruption in the pool allocator. I'm not sure, though. Would it be possible to delay creation beforehand in `declareVariable()`?

File: SPIRV/GlslangToSpv.cpp [hunk 4]
```diff
@@ -3138,6 +3139,7 @@ bool TGlslangToSpvTraverser::visitAggregate(glslang::TVisit visit, glslang::TInt
                 currentFunction->setDebugLineInfo(sourceFileId, loc.line, loc.column);
             }
         } else {
+            currentFunction = nullptr;
```
Comment (dnovillo) [inline SPIRV/GlslangToSpv.cpp]: This is in the middle of processing `currentFunction`, could you add a comment on why this is needed? Won't this confuse downstream processing?

File: glslang/Include/intermediate.h [hunk 5]
```diff
@@ -1092,6 +1095,37 @@ struct TIntermNodePair {
...
+    const TIntermVariableDecl* getAsVariableDecl() const override { return this; }
+
+    TIntermSymbol* getDeclSymbol() { return declSymbol; }
+    const TIntermSymbol* getDeclSymbol() const { return declSymbol; }
+
+    TIntermNode* getInitNode() { return initNode; }
+    const TIntermNode* getInitNode() const { return initNode; }
+
+private:
+    // This symbol is an imaginary access of the variable defined, which isn't an AST node and
```
Comment (dnovillo) [inline glslang/Include/intermediate.h]: Maybe substitute "This symbol is an imaginary access" with something like "This symbol represents the declared variable at its declaration point. It's not traversed by default. To traverse it, the visitor needs to have `includeDeclSymbol` enabled." Also, add some comment on what type of traversals will need this and why (I suppose this is only a debug info requirement for now).

File: SPIRV/GlslangToSpv.cpp (hunk 4 above)
Comment (qingyuanzNV) [reply SPIRV/GlslangToSpv.cpp]: This is actually quite natural here to reset `currentFunction` to `nullptr` as we are at the end of processing a function definition (post-visit). This handles the following case:
```
void foo() {}

int x; // <- we should have currentFunction to be nullptr while visiting AST of this

void bar() {}
```

Adding this reset logic allows us to quickly identify if we are in a local scope.

Comment (dnovillo) [reply SPIRV/GlslangToSpv.cpp]: Thanks. Please add a comment here explaining this. It tripped me at first :)

File: glslang/Include/intermediate.h (hunk 5 above)
Comment (qingyuanzNV) [reply glslang/Include/intermediate.h]: Done

File: glslang/MachineIndependent/glslang.y (hunk 3 above)
Comment (qingyuanzNV) [reply glslang/MachineIndependent/glslang.y]: Instead of freeing this node (this was old workaround as this couldn't be done easily), I figured out an approach to store the declaration node within `TIntermLoop`. Please help review.

**Review by dnovillo — APPROVED** [review]
LGTM. Thanks for all the fixes.
